Support `llvm.umin.*` and related intrinsics

LLVM has a family of four intrinsics that return the minimum or maximum of two values, using either signed or unsigned comparisons:

* [`llvm.smax.*`](https://releases.llvm.org/14.0.0/docs/LangRef.html#llvm-smax-intrinsic)
* [`llvm.smin.*`](https://releases.llvm.org/14.0.0/docs/LangRef.html#llvm-smin-intrinsic)
* [`llvm.umax.*`](https://releases.llvm.org/14.0.0/docs/LangRef.html#llvm-umax-intrinsic)
* [`llvm.umin.*`](https://releases.llvm.org/14.0.0/docs/LangRef.html#llvm-umin-intrinsic)

While these intrinsics were introduced in LLVM 12, one is much more likely to encounter them in practice in LLVM 13 or later. For example, this program produces an `llvm.umin.i32` intrinsic when compiled with optimization on Clang 13.0.0:

```c
#include <stddef.h>
#include <stdint.h>
#include <stdlib.h>
#include <crucible.h>

#define MIN_LEN 5

uint32_t *glob;

void f(uint32_t *x, uint32_t length) {
  for (uint32_t i = 0; i < length && i < MIN_LEN; i++) {
    x[i] = i;
  }
}

int main() {
  uint32_t length = crucible_uint32_t("length");
  assuming(length >= MIN_LEN);
  glob = malloc(sizeof(uint32_t) * length);
  f(glob, length);
  return 0;
}
```

While `crux-llvm` is able to verify this program with Clang 10.0.0, it fails to verify it with Clang 13.0.0:

```
$ PATH=~/Software/clang+llvm-13.0.0/bin:$PATH ~/Software/crux-llvm-0.5/bin/crux-llvm test.c
[Crux] Using pointer width: 64 for file crux-build/crux~test.bc
[Crux] Simulating function main
[Crux] Attempting to prove verification conditions.
[Crux] *** debug executable: results/test/debug-11
[Crux] *** break on line: 11
[Crux] Found counterexample for verification goal
[Crux]   test.c:11:3: error: in main
[Crux]   Failed to load function handle
[Crux]   Details:
[Crux]     No implementation or override found for pointer: "llvm.umin.i32"
[Crux]     The given pointer could not be resolved to a callable function
[Crux]     No implementation or override found for pointer: "llvm.umin.i32"
[Crux]     Attempting to load callable function with type: i32(i32, i32)
[Crux]       Via pointer: Global symbol "llvm.umin.i32" (2, 0x0:[64])
[Crux]     In memory state:
[Crux]       Branch frame
[Crux]         No writes or allocations
[Crux]       Stack frame main
[Crux]         Allocations:
[Crux]           HeapAlloc 11 0x200000000:[64] Mutable 16-byte-aligned <malloc> test.c:19:10
[Crux]         Writes:
[Crux]           Indexed chunk:
[Crux]             10 |->   *(10, 0x0:[64]) := (11, 0x0:[64])
[Crux]       Base memory
[Crux]         Allocations:
[Crux]           GlobalAlloc 10 0x8:[64] Mutable 8-byte-aligned [global variable  ] glob
[Crux]           GlobalAlloc 9 0x7:[64] Immutable 1-byte-aligned [global variable  ] .str.1
[Crux]           GlobalAlloc 8 0x7:[64] Immutable 1-byte-aligned [global variable  ] .str
[Crux]           GlobalAlloc 7 0x0:[64] Immutable 1-byte-aligned [defined function ] main
[Crux]           GlobalAlloc 6 0x0:[64] Immutable 1-byte-aligned [defined function ] f
[Crux]           GlobalAlloc 5 0x0:[64] Immutable 1-byte-aligned [external function] malloc
[Crux]           GlobalAlloc 4 0x0:[64] Immutable 1-byte-aligned [external function] crucible_assume
[Crux]           GlobalAlloc 3 0x0:[64] Immutable 1-byte-aligned [external function] crucible_int32_t
[Crux]           GlobalAlloc 2 0x0:[64] Immutable 1-byte-aligned [external function] llvm.umin.i32
[Crux]           GlobalAlloc 1 0x0:[64] Immutable 1-byte-aligned [external function] llvm.dbg.value
[Crux]         Writes:
[Crux]           Indexed chunk:
[Crux]             8 |->   *(8, 0x0:[64]) := "length\NUL"
[Crux]             9 |->   *(9, 0x0:[64]) := "test.c\NUL"
[Crux]             10 |->   *(10, 0x0:[64]) := 0
[Crux] Goal status:
[Crux]   Total: 3
[Crux]   Proved: 2
[Crux]   Disproved: 1
[Crux]   Incomplete: 0
[Crux]   Unknown: 0
[Crux] Overall status: Invalid.
```

Here is the corresponding bitcode for the `f` function:

```
; Function Attrs: nofree norecurse nosync nounwind uwtable writeonly
define dso_local void @f(i32* nocapture %0, i32 %1) local_unnamed_addr #0 !dbg !14 {
  call void @llvm.dbg.value(metadata i32* %0, metadata !19, metadata !DIExpression()), !dbg !23
  call void @llvm.dbg.value(metadata i32 %1, metadata !20, metadata !DIExpression()), !dbg !23
  call void @llvm.dbg.value(metadata i32 0, metadata !21, metadata !DIExpression()), !dbg !24
  %3 = icmp eq i32 %1, 0, !dbg !25
  br i1 %3, label %9, label %4, !dbg !27

4:                                                ; preds = %2
  %5 = add i32 %1, -1, !dbg !27
  %6 = call i32 @llvm.umin.i32(i32 %5, i32 4), !dbg !27
  %7 = add nuw nsw i32 %6, 1, !dbg !27
  %8 = zext i32 %7 to i64, !dbg !28
  br label %10, !dbg !27

9:                                                ; preds = %10, %2
  ret void, !dbg !29

10:                                               ; preds = %4, %10
  %11 = phi i64 [ 0, %4 ], [ %14, %10 ]
  call void @llvm.dbg.value(metadata i64 %11, metadata !21, metadata !DIExpression()), !dbg !24
  %12 = getelementptr inbounds i32, i32* %0, i64 %11, !dbg !30
  %13 = trunc i64 %11 to i32, !dbg !32
  store i32 %13, i32* %12, align 4, !dbg !32, !tbaa !33
  %14 = add nuw nsw i64 %11, 1, !dbg !37
  call void @llvm.dbg.value(metadata i64 %14, metadata !21, metadata !DIExpression()), !dbg !24
  %15 = icmp eq i64 %14, %8, !dbg !28
  br i1 %15, label %9, label %10, !dbg !27, !llvm.loop !38
}
```

It should be relatively straightforward to add support for `llvm.umin.*` and `llvm.umax.*`. by using `what4`'s `bvUlt` and `bvUgt` functions to implement them, and similarly for `llvm.smin.*` and `llvm.smax.*`.

--- a/crucible-llvm/src/Lang/Crucible/LLVM/Intrinsics.hs
+++ b/crucible-llvm/src/Lang/Crucible/LLVM/Intrinsics.hs
@@ -232,6 +232,15 @@ declare_overrides =
   , polymorphic1_llvm_override "llvm.umul.with.overflow"
       (\w -> SomeLLVMOverride (LLVM.llvmUmulWithOverflow w))
 
+  , polymorphic1_llvm_override "llvm.smax"
+      (\w -> SomeLLVMOverride (LLVM.llvmSmax w))
+  , polymorphic1_llvm_override "llvm.smin"
+      (\w -> SomeLLVMOverride (LLVM.llvmSmin w))
+  , polymorphic1_llvm_override "llvm.umax"
+      (\w -> SomeLLVMOverride (LLVM.llvmUmax w))
+  , polymorphic1_llvm_override "llvm.umin"
+      (\w -> SomeLLVMOverride (LLVM.llvmUmin w))
+
   , basic_llvm_override LLVM.llvmCopysignOverride_F32
   , basic_llvm_override LLVM.llvmCopysignOverride_F64
   , basic_llvm_override LLVM.llvmFabsF32

--- a/crucible-llvm/src/Lang/Crucible/LLVM/Intrinsics/LLVM.hs
+++ b/crucible-llvm/src/Lang/Crucible/LLVM/Intrinsics/LLVM.hs
@@ -503,6 +503,50 @@ llvmUmulWithOverflow w =
   [llvmOvr| { #w, i1 } $nm ( #w, #w ) |]
   (\memOps bak args -> Ctx.uncurryAssignment (callUmulWithOverflow bak memOps) args)
 
+llvmUmax ::
+  (1 <= w, IsSymInterface sym) =>
+  NatRepr w ->
+  LLVMOverride p sym
+     (EmptyCtx ::> BVType w ::> BVType w)
+     (BVType w)
+llvmUmax w =
+  let nm = L.Symbol ("llvm.umax.i" ++ show (natValue w)) in
+    [llvmOvr| #w $nm( #w, #w ) |]
+    (\memOps bak args -> Ctx.uncurryAssignment (callUmax bak memOps) args)
+
+llvmUmin ::
+  (1 <= w, IsSymInterface sym) =>
+  NatRepr w ->
+  LLVMOverride p sym
+     (EmptyCtx ::> BVType w ::> BVType w)
+     (BVType w)
+llvmUmin w =
+  let nm = L.Symbol ("llvm.umin.i" ++ show (natValue w)) in
+    [llvmOvr| #w $nm( #w, #w ) |]
+    (\memOps bak args -> Ctx.uncurryAssignment (callUmin bak memOps) args)
+
+llvmSmax ::
+  (1 <= w, IsSymInterface sym) =>
+  NatRepr w ->
+  LLVMOverride p sym
+     (EmptyCtx ::> BVType w ::> BVType w)
+     (BVType w)
+llvmSmax w =
+  let nm = L.Symbol ("llvm.smax.i" ++ show (natValue w)) in
+    [llvmOvr| #w $nm( #w, #w ) |]
+    (\memOps bak args -> Ctx.uncurryAssignment (callSmax bak memOps) args)
+
+llvmSmin ::
+  (1 <= w, IsSymInterface sym) =>
+  NatRepr w ->
+  LLVMOverride p sym
+     (EmptyCtx ::> BVType w ::> BVType w)
+     (BVType w)
+llvmSmin w =
+  let nm = L.Symbol ("llvm.smin.i" ++ show (natValue w)) in
+    [llvmOvr| #w $nm( #w, #w ) |]
+    (\memOps bak args -> Ctx.uncurryAssignment (callSmin bak memOps) args)
+
 llvmCtlz
   :: (1 <= w, IsSymInterface sym)
   => NatRepr w ->
@@ -1144,6 +1188,54 @@ callUmulWithOverflow bak _mvar
        (ov, z) <- mulUnsignedOF sym x y
        ov' <- predToBV sym ov (knownNat @1)
        return (Empty :> RV z :> RV ov')
+
+callUmax
+  :: (1 <= w, IsSymBackend sym bak)
+  => bak
+  -> GlobalVar Mem
+  -> RegEntry sym (BVType w)
+  -> RegEntry sym (BVType w)
+  -> OverrideSim p sym ext r args ret (RegValue sym (BVType w))
+callUmax bak _mvar (regValue -> x) (regValue -> y) = liftIO $
+  do let sym = backendGetSym bak
+     xGtY <- bvUgt sym x y
+     bvIte sym xGtY x y
+
+callUmin
+  :: (1 <= w, IsSymBackend sym bak)
+  => bak
+  -> GlobalVar Mem
+  -> RegEntry sym (BVType w)
+  -> RegEntry sym (BVType w)
+  -> OverrideSim p sym ext r args ret (RegValue sym (BVType w))
+callUmin bak _mvar (regValue -> x) (regValue -> y) = liftIO $
+  do let sym = backendGetSym bak
+     xLtY <- bvUlt sym x y
+     bvIte sym xLtY x y
+
+callSmax
+  :: (1 <= w, IsSymBackend sym bak)
+  => bak
+  -> GlobalVar Mem
+  -> RegEntry sym (BVType w)
+  -> RegEntry sym (BVType w)
+  -> OverrideSim p sym ext r args ret (RegValue sym (BVType w))
+callSmax bak _mvar (regValue -> x) (regValue -> y) = liftIO $
+  do let sym = backendGetSym bak
+     xGtY <- bvSgt sym x y
+     bvIte sym xGtY x y
+
+callSmin
+  :: (1 <= w, IsSymBackend sym bak)
+  => bak
+  -> GlobalVar Mem
+  -> RegEntry sym (BVType w)
+  -> RegEntry sym (BVType w)
+  -> OverrideSim p sym ext r args ret (RegValue sym (BVType w))
+callSmin bak _mvar (regValue -> x) (regValue -> y) = liftIO $
+  do let sym = backendGetSym bak
+     xLtY <- bvSlt sym x y
+     bvIte sym xLtY x y
 
 
 callCttz

--- a/crux-llvm/test-data/golden/golden/T974.at-least-clang12.z3.good
+++ b/crux-llvm/test-data/golden/golden/T974.at-least-clang12.z3.good
@@ -1,0 +1,1 @@
+[Crux] Overall status: Valid.

--- a/crux-llvm/test-data/golden/golden/T974.c.orig
+++ b/crux-llvm/test-data/golden/golden/T974.c.orig
@@ -1,0 +1,58 @@
+// Unit tests for `llvm.smin.*` and related intrinsics. Annoyingly, I don't know
+// of a way to make Clang compile `smin` and the related functions below into
+// the corresponding LLVM intrinsics, so some manual .ll edits are required.
+
+#include <crucible.h>
+#include <stdint.h>
+
+int32_t smin(int32_t x, int32_t y) {
+  return (x < y) ? x : y;
+}
+
+int32_t smax(int32_t x, int32_t y) {
+  return (x > y) ? x : y;
+}
+
+uint32_t umin(uint32_t x, uint32_t y) {
+  return (x < y) ? x : y;
+}
+
+uint32_t umax(uint32_t x, uint32_t y) {
+  return (x > y) ? x : y;
+}
+
+int main(void) {
+  // smin
+  check(smin(2, 3) == 2);
+  check(smin(3, 2) == 2);
+  int32_t smin1 = crucible_int32_t("smin1");
+  int32_t smin2 = crucible_int32_t("smin2");
+  assuming(smin1 <= smin2);
+  check(smin(smin1, smin2) == smin1);
+
+  // smax
+  check(smax(2, 3) == 3);
+  check(smax(3, 2) == 3);
+  int32_t smax1 = crucible_int32_t("smax1");
+  int32_t smax2 = crucible_int32_t("smax2");
+  assuming(smax1 >= smax2);
+  check(smax(smax1, smax2) == smax1);
+
+  // umin
+  check(umin(2, 3) == 2);
+  check(umin(3, 2) == 2);
+  uint32_t umin1 = crucible_uint32_t("umin1");
+  uint32_t umin2 = crucible_uint32_t("umin2");
+  assuming(umin1 <= umin2);
+  check(umin(umin1, umin2) == umin1);
+
+  // umax
+  check(umax(2, 3) == 3);
+  check(umax(3, 2) == 3);
+  uint32_t umax1 = crucible_uint32_t("umax1");
+  uint32_t umax2 = crucible_uint32_t("umax2");
+  assuming(umax1 >= umax2);
+  check(umax(umax1, umax2) == umax1);
+
+  return 0;
+}

--- a/crux-llvm/test-data/golden/golden/T974.ll
+++ b/crux-llvm/test-data/golden/golden/T974.ll
@@ -1,0 +1,148 @@
+; ModuleID = 'T974.c'
+source_filename = "T974.c"
+target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
+target triple = "x86_64-unknown-linux-gnu"
+
+@.str = private unnamed_addr constant [7 x i8] c"T974.c\00", align 1
+@.str.1 = private unnamed_addr constant [6 x i8] c"smin1\00", align 1
+@.str.2 = private unnamed_addr constant [6 x i8] c"smin2\00", align 1
+@.str.3 = private unnamed_addr constant [6 x i8] c"smax1\00", align 1
+@.str.4 = private unnamed_addr constant [6 x i8] c"smax2\00", align 1
+@.str.5 = private unnamed_addr constant [6 x i8] c"umin1\00", align 1
+@.str.6 = private unnamed_addr constant [6 x i8] c"umin2\00", align 1
+@.str.7 = private unnamed_addr constant [6 x i8] c"umax1\00", align 1
+@.str.8 = private unnamed_addr constant [6 x i8] c"umax2\00", align 1
+
+declare i32 @llvm.smin.i32(i32, i32)
+declare i32 @llvm.smax.i32(i32, i32)
+declare i32 @llvm.umin.i32(i32, i32)
+declare i32 @llvm.umax.i32(i32, i32)
+
+; Function Attrs: norecurse nounwind readnone uwtable willreturn
+define dso_local i32 @smin(i32 %x, i32 %y) local_unnamed_addr #0 {
+entry:
+  ; %cmp = icmp slt i32 %x, %y
+  ; %cond = select i1 %cmp, i32 %x, i32 %y
+  %cond = call i32 @llvm.smin.i32(i32 %x, i32 %y)
+  ret i32 %cond
+}
+
+; Function Attrs: norecurse nounwind readnone uwtable willreturn
+define dso_local i32 @smax(i32 %x, i32 %y) local_unnamed_addr #0 {
+entry:
+  ; %cmp = icmp sgt i32 %x, %y
+  ; %cond = select i1 %cmp, i32 %x, i32 %y
+  %cond = call i32 @llvm.smax.i32(i32 %x, i32 %y)
+  ret i32 %cond
+}
+
+; Function Attrs: norecurse nounwind readnone uwtable willreturn
+define dso_local i32 @umin(i32 %x, i32 %y) local_unnamed_addr #0 {
+entry:
+  ; %cmp = icmp ult i32 %x, %y
+  ; %cond = select i1 %cmp, i32 %x, i32 %y
+  %cond = call i32 @llvm.umin.i32(i32 %x, i32 %y)
+  ret i32 %cond
+}
+
+; Function Attrs: norecurse nounwind readnone uwtable willreturn
+define dso_local i32 @umax(i32 %x, i32 %y) local_unnamed_addr #0 {
+entry:
+  ; %cmp = icmp ugt i32 %x, %y
+  ; %cond = select i1 %cmp, i32 %x, i32 %y
+  %cond = call i32 @llvm.umax.i32(i32 %x, i32 %y)
+  ret i32 %cond
+}
+
+; Function Attrs: nounwind uwtable
+define dso_local i32 @main() local_unnamed_addr #1 {
+entry:
+  %call = call i32 @smin(i32 2, i32 3)
+  %cmp = icmp eq i32 %call, 2
+  %conv1 = zext i1 %cmp to i8
+  call void @crucible_assert(i8 zeroext %conv1, i8* getelementptr inbounds ([7 x i8], [7 x i8]* @.str, i64 0, i64 0), i32 26) #3
+  %call2 = call i32 @smin(i32 3, i32 2)
+  %cmp3 = icmp eq i32 %call2, 2
+  %conv5 = zext i1 %cmp3 to i8
+  call void @crucible_assert(i8 zeroext %conv5, i8* getelementptr inbounds ([7 x i8], [7 x i8]* @.str, i64 0, i64 0), i32 27) #3
+  %call6 = call i32 @crucible_int32_t(i8* getelementptr inbounds ([6 x i8], [6 x i8]* @.str.1, i64 0, i64 0)) #3
+  %call7 = call i32 @crucible_int32_t(i8* getelementptr inbounds ([6 x i8], [6 x i8]* @.str.2, i64 0, i64 0)) #3
+  %cmp8 = icmp sle i32 %call6, %call7
+  %conv10 = zext i1 %cmp8 to i8
+  call void @crucible_assume(i8 zeroext %conv10, i8* getelementptr inbounds ([7 x i8], [7 x i8]* @.str, i64 0, i64 0), i32 30) #3
+  %call11 = call i32 @smin(i32 %call6, i32 %call7)
+  %cmp12 = icmp eq i32 %call11, %call6
+  %conv14 = zext i1 %cmp12 to i8
+  call void @crucible_assert(i8 zeroext %conv14, i8* getelementptr inbounds ([7 x i8], [7 x i8]* @.str, i64 0, i64 0), i32 31) #3
+  %call15 = call i32 @smax(i32 2, i32 3)
+  %cmp16 = icmp eq i32 %call15, 3
+  %conv18 = zext i1 %cmp16 to i8
+  call void @crucible_assert(i8 zeroext %conv18, i8* getelementptr inbounds ([7 x i8], [7 x i8]* @.str, i64 0, i64 0), i32 34) #3
+  %call19 = call i32 @smax(i32 3, i32 2)
+  %cmp20 = icmp eq i32 %call19, 3
+  %conv22 = zext i1 %cmp20 to i8
+  call void @crucible_assert(i8 zeroext %conv22, i8* getelementptr inbounds ([7 x i8], [7 x i8]* @.str, i64 0, i64 0), i32 35) #3
+  %call23 = call i32 @crucible_int32_t(i8* getelementptr inbounds ([6 x i8], [6 x i8]* @.str.3, i64 0, i64 0)) #3
+  %call24 = call i32 @crucible_int32_t(i8* getelementptr inbounds ([6 x i8], [6 x i8]* @.str.4, i64 0, i64 0)) #3
+  %cmp25 = icmp sge i32 %call23, %call24
+  %conv27 = zext i1 %cmp25 to i8
+  call void @crucible_assume(i8 zeroext %conv27, i8* getelementptr inbounds ([7 x i8], [7 x i8]* @.str, i64 0, i64 0), i32 38) #3
+  %call28 = call i32 @smax(i32 %call23, i32 %call24)
+  %cmp29 = icmp eq i32 %call28, %call23
+  %conv31 = zext i1 %cmp29 to i8
+  call void @crucible_assert(i8 zeroext %conv31, i8* getelementptr inbounds ([7 x i8], [7 x i8]* @.str, i64 0, i64 0), i32 39) #3
+  %call32 = call i32 @umin(i32 2, i32 3)
+  %cmp33 = icmp eq i32 %call32, 2
+  %conv35 = zext i1 %cmp33 to i8
+  call void @crucible_assert(i8 zeroext %conv35, i8* getelementptr inbounds ([7 x i8], [7 x i8]* @.str, i64 0, i64 0), i32 42) #3
+  %call36 = call i32 @umin(i32 3, i32 2)
+  %cmp37 = icmp eq i32 %call36, 2
+  %conv39 = zext i1 %cmp37 to i8
+  call void @crucible_assert(i8 zeroext %conv39, i8* getelementptr inbounds ([7 x i8], [7 x i8]* @.str, i64 0, i64 0), i32 43) #3
+  %call40 = call i32 @crucible_int32_t(i8* getelementptr inbounds ([6 x i8], [6 x i8]* @.str.5, i64 0, i64 0)) #3
+  %call41 = call i32 @crucible_int32_t(i8* getelementptr inbounds ([6 x i8], [6 x i8]* @.str.6, i64 0, i64 0)) #3
+  %cmp42 = icmp ule i32 %call40, %call41
+  %conv44 = zext i1 %cmp42 to i8
+  call void @crucible_assume(i8 zeroext %conv44, i8* getelementptr inbounds ([7 x i8], [7 x i8]* @.str, i64 0, i64 0), i32 46) #3
+  %call45 = call i32 @umin(i32 %call40, i32 %call41)
+  %cmp46 = icmp eq i32 %call45, %call40
+  %conv48 = zext i1 %cmp46 to i8
+  call void @crucible_assert(i8 zeroext %conv48, i8* getelementptr inbounds ([7 x i8], [7 x i8]* @.str, i64 0, i64 0), i32 47) #3
+  %call49 = call i32 @umax(i32 2, i32 3)
+  %cmp50 = icmp eq i32 %call49, 3
+  %conv52 = zext i1 %cmp50 to i8
+  call void @crucible_assert(i8 zeroext %conv52, i8* getelementptr inbounds ([7 x i8], [7 x i8]* @.str, i64 0, i64 0), i32 50) #3
+  %call53 = call i32 @umax(i32 3, i32 2)
+  %cmp54 = icmp eq i32 %call53, 3
+  %conv56 = zext i1 %cmp54 to i8
+  call void @crucible_assert(i8 zeroext %conv56, i8* getelementptr inbounds ([7 x i8], [7 x i8]* @.str, i64 0, i64 0), i32 51) #3
+  %call57 = call i32 @crucible_int32_t(i8* getelementptr inbounds ([6 x i8], [6 x i8]* @.str.7, i64 0, i64 0)) #3
+  %call58 = call i32 @crucible_int32_t(i8* getelementptr inbounds ([6 x i8], [6 x i8]* @.str.8, i64 0, i64 0)) #3
+  %cmp59 = icmp uge i32 %call57, %call58
+  %conv61 = zext i1 %cmp59 to i8
+  call void @crucible_assume(i8 zeroext %conv61, i8* getelementptr inbounds ([7 x i8], [7 x i8]* @.str, i64 0, i64 0), i32 54) #3
+  %call62 = call i32 @umax(i32 %call57, i32 %call58)
+  %cmp63 = icmp eq i32 %call62, %call57
+  %conv65 = zext i1 %cmp63 to i8
+  call void @crucible_assert(i8 zeroext %conv65, i8* getelementptr inbounds ([7 x i8], [7 x i8]* @.str, i64 0, i64 0), i32 55) #3
+  ret i32 0
+}
+
+declare dso_local void @crucible_assert(i8 zeroext, i8*, i32) local_unnamed_addr #2
+
+declare dso_local i32 @crucible_int32_t(i8*) local_unnamed_addr #2
+
+declare dso_local void @crucible_assume(i8 zeroext, i8*, i32) local_unnamed_addr #2
+
+attributes #0 = { norecurse nounwind readnone uwtable willreturn "disable-tail-calls"="false" "frame-pointer"="none" "less-precise-fpmad"="false" "min-legal-vector-width"="0" "no-infs-fp-math"="false" "no-jump-tables"="false" "no-nans-fp-math"="false" "no-signed-zeros-fp-math"="false" "no-trapping-math"="true" "stack-protector-buffer-size"="8" "target-cpu"="x86-64" "target-features"="+cx8,+fxsr,+mmx,+sse,+sse2,+x87" "tune-cpu"="generic" "unsafe-fp-math"="false" "use-soft-float"="false" }
+attributes #1 = { nounwind uwtable "disable-tail-calls"="false" "frame-pointer"="none" "less-precise-fpmad"="false" "min-legal-vector-width"="0" "no-infs-fp-math"="false" "no-jump-tables"="false" "no-nans-fp-math"="false" "no-signed-zeros-fp-math"="false" "no-trapping-math"="true" "stack-protector-buffer-size"="8" "target-cpu"="x86-64" "target-features"="+cx8,+fxsr,+mmx,+sse,+sse2,+x87" "tune-cpu"="generic" "unsafe-fp-math"="false" "use-soft-float"="false" }
+attributes #2 = { "disable-tail-calls"="false" "frame-pointer"="none" "less-precise-fpmad"="false" "no-infs-fp-math"="false" "no-nans-fp-math"="false" "no-signed-zeros-fp-math"="false" "no-trapping-math"="true" "stack-protector-buffer-size"="8" "target-cpu"="x86-64" "target-features"="+cx8,+fxsr,+mmx,+sse,+sse2,+x87" "tune-cpu"="generic" "unsafe-fp-math"="false" "use-soft-float"="false" }
+attributes #3 = { nounwind }
+
+!llvm.module.flags = !{!0}
+!llvm.ident = !{!1}
+!llvm.commandline = !{!2}
+
+!0 = !{i32 1, !"wchar_size", i32 4}
+!1 = !{!"clang version 12.0.1"}
+!2 = !{!"/home/rscott/Software/clang+llvm-12.0.1/bin/clang-12 -fno-discard-value-names -frecord-command-line -S -D CRUCIBLE -emit-llvm -I ../../../c-src/includes -O1 T974.c -o T974.ll"}

